vpr/src/place/move_utils.cpp

```diff
@@ -1299,3 +1290,148 @@ std::string e_move_result_to_string(e_move_result move_outcome) {
     std::string move_result_to_string[] = {"Rejected", "Accepted", "Aborted"};
     return move_result_to_string[move_outcome];
 }
+
+int find_free_layer(t_logical_block_type_ptr logical_block, const t_pl_loc& loc) {
+    const auto& device_ctx = g_vpr_ctx.device();
+    const auto& place_ctx = g_vpr_ctx.placement();
+
+    // TODO: Compatible layer vector should be shuffled first, and then iterated through
+    int free_layer = loc.layer;
+    VTR_ASSERT(loc.layer != OPEN);
+    if (device_ctx.grid.get_num_layers() > 1) {
+        const auto& compatible_layers = place_ctx.compressed_block_grids[logical_block->index].get_layer_nums();
+        if (compatible_layers.size() > 1) {
+            if (place_ctx.grid_blocks.block_at_location(loc) != EMPTY_BLOCK_ID) {
+                for (const auto& layer : compatible_layers) {
+                    if (layer != free_layer) {
+                        if (place_ctx.grid_blocks.block_at_location(loc) == EMPTY_BLOCK_ID) {
+                            free_layer = layer;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return free_layer;
+}
+
+int get_random_layer(t_logical_block_type_ptr logical_block) {
+    const auto& compatible_layers = g_vpr_ctx.placement().compressed_block_grids[logical_block->index].get_layer_nums();
+    VTR_ASSERT(!compatible_layers.empty());
+    int layer_num = OPEN;
+    if (compatible_layers.size() == 1) {
+        layer_num = compatible_layers[0];
+    } else {
+        layer_num = compatible_layers[vtr::irand(compatible_layers.size() - 1)];
+    }
+
+    return layer_num;
+}
+
+t_bb union_2d_bb(const std::vector<t_2D_bb>& bb_vec) {
+    t_bb merged_bb;
+
```

Comment:
Standard technique: 
assign first entry to xmin, xmax, etc. and then loop over rest doing min and max.

Reply:
Or comment if there is some assumption / encoding of invalid bounding boxes here.